File: exams/bütünleme1/run1.m
clear;clc;

z=sym('z');
kp=sym('kp','real');
kd=sym('kd','real');

Gzz=1/(z+1);
Fzz=(kd*z+kp)/z;
Tzz=(Fzz*Gzz)/(1+Fzz*Gzz);
[nz,dz]=numden(Tzz);
coef_num=coeffs(nz,z,'all');
coef_den=coeffs(dz,z,'all');
coef_num=coef_num/coef_den(1);
coef_den=coef_den/coef_den(1);

disp("num:");disp(coef_num);
disp("den:");disp(coef_den);

ts=2;
os=16.3/100;

zeta=-log(os)/sqrt(pi^2+log(os)^2);
wn=4/(ts*zeta);
zeta=round(zeta,2);
wn=round(wn,2);

disp("zeta:"+string(zeta));
disp("wn:"+string(wn));
sval=-zeta*wn+1i*sqrt(1-zeta^2)*wn;
disp("s:"+string(sval));

T=1;
zval=exp(sval*T);
disp("z:"+string(zval));

pdz=(z-zval)*(z-conj(zval));
pdz=collect(pdz,z);
disp("p(z):");
disp(vpa(pdz,4));

prob=coeffs(pdz,z,'all')==coef_den;
disp("prob:");
for i=1:length(prob)
    disp(vpa(prob(i),4));
end
sol=solve(prob);
kpval=double(sol.kp);
kdval=double(sol.kd);
disp("kp:"+string(kpval));
disp("kd:"+string(kdval));

Fz=tf([kdval kpval],[1 0],T);
Gz=tf(1,[1 1],T);
Tz=feedback(Fz*Gz,1);
Fz
zpk(Fz)
Tz
zpk(Tz)

File: lecture notes/python/lec8/run1.py
import matplotlib.pyplot as plt
import numpy as np
import sympy as sym
import tbcontrol
import control
############################
# q1
############################
def fun_q1():
    T=0.2
    z,kd,kp=sym.symbols("z,kd,kp")
    num=0.1648
    den=z-0.6703
    Gz=control.tf(num,np.array(sym.Poly(den,z).all_coeffs(),dtype=np.float64),T)

    os=10/100
    ts=1
    zeta=-np.log(os)/np.sqrt(np.pi**2+np.log(os)**2)
    wn=4/(ts*zeta)
    
    pole_s=-zeta*wn+1j*wn*np.sqrt(1-zeta**2)
    pole_z=np.exp(pole_s*T)
    pdz=(z-pole_z)*(z-np.conj(pole_z))
    pdz=sym.expand(pdz)

    Gzz=num/den
    Fzz=((kd+kp)*z-kd)/z

    Tzz=(Fzz*Gzz)/(1+Fzz*Gzz)
    Tzz=sym.simplify(Tzz)
    pzz,pcz=sym.fraction(Tzz)
    pcz=sym.expand(pcz)

    coef_pcz=sym.Poly(pcz,z).all_coeffs()
    coef_pdz=sym.Poly(pdz,z).all_coeffs()
    prob=[]
    for i in range(1,len(coef_pcz)):
        prob.append(coef_pcz[i]-coef_pdz[i])
    sol=sym.solve(prob)
    print(sol)

    kdval=sol[kd]
    kpval=sol[kp]
    kdval=float(kdval)
    kpval=float(kpval)
    Fz=control.tf([kdval+kpval,-kdval],np.array([1,0]),T)
    Tz=control.feedback(control.series(Fz,Gz),1)
    
    t, y = control.step_response(Tz)

    plt.stem(t,y)
    plt.xlim([0,4])
    plt.title("Basamak Yanıtı")
    plt.xlabel("Zaman(s)")
    plt.show()
    
fun_q1()